Relocate reserveNTrampolines to x86 CodeGenerator

0xdaryl opened the pull request:
Move the `reserveNTrampolines()` implementation from the FrontEnd to
the x86 CodeGenerator where it more naturally belongs.  The API is
also simplified to only accept the number of trampolines to reserve.
The rest of the original API can be derived from other queries
on the `TR::CodeGenerator` object.

Nothing uses this function yet.  A subsequent commit will delete the
implementation in the FrontEnd and re-route all references to the
version in CodeGenerator.

Signed-off-by: Daryl Maier <maier@ca.ibm.com>

0xdaryl commented:
I'd appreciate a review+merge from at least one of @andrewcraik @ymanton @fjeremic .

This PR essentially copies the `reserveNTrampolines` implementation from the FrontEnd into CodeGenerator.  Subsequent commits will remove it from FE and replace references to the CodeGenerator version.  Some pre-req OMR work is needed before that happens though.

fjeremic commented on runtime/tr.source/trj9/x/codegen/J9CodeGenerator.cpp:
```diff
@@ -398,3 +402,71 @@ J9::X86::CodeGenerator::suppressInliningOfRecognizedMethod(TR::RecognizedMethod
+   TR::CodeCache *newCache = curCache;
+   OMR::CodeCacheErrorCode::ErrorCode status = OMR::CodeCacheErrorCode::ERRORCODE_SUCCESS;
+
+   TR_ASSERT(curCache->isReserved(), "assertion failure");
```
This assertion message is not very useful. While we are in this area of code could we improve the message?

fjeremic commented on runtime/tr.source/trj9/x/codegen/J9CodeGenerator.cpp:
```diff
@@ -398,3 +402,71 @@ J9::X86::CodeGenerator::suppressInliningOfRecognizedMethod(TR::RecognizedMethod
+
+   if (!fej9->isAOT_DEPRECATED_DO_NOT_USE())
+      {
+      status = curCache->reserveNTrampolines(numTrampolines);
```
Should the `J9::X86::CodeGenerator::reserveNTrampolines` also return such a status code? Would anyone care about the result of `J9::X86::CodeGenerator::reserveNTrampolines`?

fjeremic commented on runtime/tr.source/trj9/x/codegen/J9CodeGenerator.cpp:
```diff
@@ -398,3 +402,71 @@ J9::X86::CodeGenerator::suppressInliningOfRecognizedMethod(TR::RecognizedMethod
+
+      // If the old CC had pre-loaded code, the current compilation may have initialized it and will therefore depend on it
+      // so we should initialize it in the new CC as well
+      // XXX: We could avoid this if we knew for sure that this compile wasn't the one who initialized it
```
Is this something we should have an issue opened for?

0xdaryl commented:
Thanks for your review and comments @fjeremic.  My intent with this PR was simply to move the code as-is from the FrontEnd class in VMJ9.cpp to the CodeGenerator class without making any changes to its logic (or even deeply understand the logic therein for that matter).

However, I do think your comments have merit.  The assert text is easy to fix and I provided a fix for that.  At the moment, the function in question expects success in order to continue.  If not, the compilation is expected to be aborted via an exception throw.  That doesn't appear to be the case on all paths through this function and is something that should be sorted out.  Your other comment about creating an issue is also warranted, but I think that should be done understanding the semantics and the context of that code.

This PR is just part of a larger set of issues to clean up the trampoline logic in general and get the code in better shape to support trampolines for native functions.  If its acceptable to you, I'll create an issue to track addressing your comments raised in this issue in more detail so that we can revisit them when we actually tackle cleaning up the trampoline logic itself.

fjeremic commented:
> This PR is just part of a larger set of issues to clean up the trampoline logic in general and get the code in better shape to support trampolines for native functions. If its acceptable to you, I'll create an issue to track addressing your comments raised in this issue in more detail so that we can revisit them when we actually tackle cleaning up the trampoline logic itself.

This is perfectly reasonable. Thanks for the explanation! I understand the context of the code move now.

fjeremic commented:
Jenkins test sanity
